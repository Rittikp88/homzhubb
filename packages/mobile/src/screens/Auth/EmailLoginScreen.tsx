import React from 'react';
import { StyleSheet, View } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';
import { IState } from '@homzhub/common/src/modules/interfaces';
import { UserActions } from '@homzhub/common/src/modules/user/actions';
import {
  IEmailLoginPayload,
  ILoginFormData,
  IOtpLoginPayload,
  LoginTypes,
} from '@homzhub/common/src/domain/repositories/interfaces';
import { theme } from '@homzhub/common/src/styles/theme';
import { icons } from '@homzhub/common/src/assets/icon';
import { DetailedHeader } from '@homzhub/common/src/components/molecules/DetailedHeader';
import { LoginForm } from '@homzhub/common/src/components/organisms/LoginForm';
import { AuthStackParamList } from '@homzhub/mobile/src/navigation/AuthStack';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';

interface IDispatchProps {
  login: (payload: IEmailLoginPayload | IOtpLoginPayload) => void;
}

type libraryProps = WithTranslation & NavigationScreenProps<AuthStackParamList, ScreensKeys.EmailLogin>;
type Props = IDispatchProps & libraryProps;

export class EmailLoginScreen extends React.PureComponent<Props> {
  public render(): React.ReactNode {
    const { t } = this.props;
    return (
      <View style={styles.container}>
        <DetailedHeader
          icon={icons.leftArrow}
          subTitle={t('auth:logInWithEmail')}
          subTitleType="large"
          subTitleColor={theme.colors.dark}
          onIconPress={this.handleIconPress}
          testID="headerIconPress"
        />
        <LoginForm
          isEmailLogin
          handleForgotPassword={this.handleForgotPassword}
          onLoginSuccess={this.handleLoginSuccess}
          testID="loginForm"
        />
      </View>
    );
  }

  private handleIconPress = (): void => {
    const { navigation } = this.props;
    navigation.goBack();
  };

  private handleForgotPassword = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.ForgotPassword);
  };

  private handleLoginSuccess = (values: ILoginFormData): void => {
    const { login } = this.props;

    const emailLoginData: IEmailLoginPayload = {
      action: LoginTypes.EMAIL,
      payload: {
        email: values.email,
        password: values.password,
      },
    };

    login(emailLoginData);
  };
}

export const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
  const { login } = UserActions;
  return bindActionCreators(
    {
      login,
    },
    dispatch
  );
};

export default connect<null, IDispatchProps, WithTranslation, IState>(
  null,
  mapDispatchToProps
)(withTranslation()(EmailLoginScreen));

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: theme.colors.white,
  },
});
