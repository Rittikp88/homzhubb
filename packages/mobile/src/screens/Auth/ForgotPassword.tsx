import React, { Component } from 'react';
import { StyleSheet, View } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { Formik, FormikProps, FormikValues } from 'formik';
import * as yup from 'yup';
import { FormUtils } from '@homzhub/common/src/utils/FormUtils';
import { AlertHelper } from '@homzhub/mobile/src/utils/AlertHelper';
import { UserService } from '@homzhub/common/src/services/UserService';
import { theme } from '@homzhub/common/src/styles/theme';
import { FormButton, FormTextInput, AnimatedHeader, Text } from '@homzhub/common/src/components';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { AuthStackParamList } from '@homzhub/mobile/src/navigation/AuthStack';

interface IForgotPasswordState {
  email: string;
}

type Props = WithTranslation & NavigationScreenProps<AuthStackParamList, ScreensKeys.ForgotPassword>;

class ForgotPassword extends Component<Props, IForgotPasswordState> {
  public state = {
    email: '',
  };

  public render(): React.ReactNode {
    const { t } = this.props;
    const formData = { ...this.state };
    return (
      <View style={styles.container}>
        <AnimatedHeader
          icon="left-arrow"
          title={t('auth:forgotPassword')}
          subTitle={t('auth:forgotPasswordDescription')}
          onIconPress={this.handleIconPress}
          headerContainerStyle={styles.headerContainer}
        />
        <View style={styles.content}>
          <Formik onSubmit={this.onSubmit} validate={FormUtils.validate(this.formSchema)} initialValues={formData}>
            {(formProps: FormikProps<FormikValues>): React.ReactElement => (
              <>
                <FormTextInput
                  formProps={formProps}
                  inputType="email"
                  name="email"
                  label="Email"
                  placeholder={t('auth:enterEmail')}
                />
                <FormButton
                  formProps={formProps}
                  type="primary"
                  title={t('auth:forgotPasswordRequestLink')}
                  // @ts-ignore
                  onPress={formProps.handleSubmit}
                  containerStyle={styles.formButtonStyle}
                />
              </>
            )}
          </Formik>
          <Text type="small" textType="semiBold" style={styles.backToLoginLink} onPress={this.navigateToLogin}>
            {t('auth:backToLogin')}
          </Text>
        </View>
      </View>
    );
  }

  private onSubmit = (formProps: IForgotPasswordState): void => {
    const { navigation } = this.props;
    const { email } = formProps;
    const payload = {
      action: 'SEND_EMAIL',
      payload: {
        email,
      },
    };
    UserService.resetPassword(payload)
      .then((response) => {
        const { email_exists, token } = response;
        if (email_exists) {
          navigation.navigate(ScreensKeys.ResetPassword, {
            token,
            email,
          });
        } else {
          AlertHelper.error({ message: 'Email does not exists' });
        }
        return null;
      })
      .catch((err) => {
        AlertHelper.error({ message: err });
      });
  };

  public navigateToLogin = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.EmailLogin);
  };

  public handleIconPress = (): void => {
    const { navigation } = this.props;
    navigation.goBack();
  };

  private formSchema = (): yup.ObjectSchema<{ email: string }> => {
    const { t } = this.props;
    return yup.object().shape({
      email: yup.string().email(t('auth:emailValidation')).required(t('auth:emailRequired')),
    });
  };
}

export default withTranslation()(ForgotPassword);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: theme.colors.white,
  },
  content: {
    paddingHorizontal: 24,
    paddingVertical: 16,
  },
  headerContainer: {
    borderBottomWidth: 0,
  },
  formButtonStyle: {
    flex: 0,
    marginVertical: 30,
  },
  backToLoginLink: {
    color: theme.colors.blue,
    alignSelf: 'center',
  },
});
