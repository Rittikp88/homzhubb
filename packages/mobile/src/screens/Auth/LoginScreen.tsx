import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';
import { withTranslation, WithTranslation } from 'react-i18next';
import { AlertHelper } from '@homzhub/mobile/src/utils/AlertHelper';
import { UserRepository } from '@homzhub/common/src/domain/repositories/UserRepository';
import { UserActions } from '@homzhub/common/src/modules/user/actions';
import { FormTextInput, AnimatedHeader, LoginForm, SocialMediaComponent } from '@homzhub/common/src/components';
import { AuthStackParamList } from '@homzhub/mobile/src/navigation/AuthStack';
import { ILoginFormData } from '@homzhub/common/src/domain/repositories/interfaces';
import { IUser } from '@homzhub/common/src/domain/models/User';
import { ISocialMediaProvider } from '@homzhub/common/src/domain/models/SocialMediaProvider';
import { IState } from '@homzhub/common/src/modules/interfaces';
import { NavigationScreenProps, OtpNavTypes, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';

interface IDispatchProps {
  loginSuccess: (data: IUser) => void;
}

interface ILoginScreenState {
  socialMediaProviders: ISocialMediaProvider[];
}

type libraryProps = WithTranslation & NavigationScreenProps<AuthStackParamList, ScreensKeys.Login>;
type Props = IDispatchProps & libraryProps;

export class LoginScreen extends Component<Props, ILoginScreenState> {
  public state = {
    socialMediaProviders: [],
  };

  public async componentDidMount(): Promise<void> {
    await this.fetchSocialMedia();
  }

  public render(): React.ReactNode {
    const { t, loginSuccess, navigation } = this.props;
    const { socialMediaProviders } = this.state;

    return (
      <AnimatedHeader
        title={t('login')}
        subTitle={t('auth:newAroundHere')}
        linkText={t('auth:signup')}
        onIconPress={this.onClosePress}
        onLinkPress={this.onSignUpClicked}
        testID="headerEvents"
      >
        <>
          <LoginForm onLoginSuccess={this.onOtpLoginPress} testID="loginForm" />
          <SocialMediaComponent
            isFromLogin
            onLoginSuccessAction={loginSuccess}
            socialMediaItems={socialMediaProviders}
            onEmailLogin={this.onEmailLoginPress}
            navigation={navigation}
            testID="socialEmailLogin"
          />
        </>
      </AnimatedHeader>
    );
  }

  private onClosePress = (): void => {
    const { navigation } = this.props;
    navigation.goBack();
  };

  private onEmailLoginPress = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.EmailLogin);
  };

  private onSignUpClicked = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.SignUp);
  };

  private onOtpLoginPress = (values: ILoginFormData, ref: () => FormTextInput | null): void => {
    const { navigation, t } = this.props;
    navigation.navigate(ScreensKeys.OTP, {
      type: OtpNavTypes.Login,
      title: t('auth:loginOtp'),
      countryCode: values.country_code,
      phone: values.phone_number,
      ref,
    });
  };

  private fetchSocialMedia = async (): Promise<void> => {
    try {
      const response = await UserRepository.getSocialMedia();
      this.setState({ socialMediaProviders: response });
    } catch (e) {
      AlertHelper.error({ message: e.message });
    }
  };
}

export const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
  const { loginSuccess } = UserActions;
  return bindActionCreators(
    {
      loginSuccess,
    },
    dispatch
  );
};

export default connect<{}, IDispatchProps, WithTranslation, IState>(
  null,
  mapDispatchToProps
)(withTranslation()(LoginScreen));
