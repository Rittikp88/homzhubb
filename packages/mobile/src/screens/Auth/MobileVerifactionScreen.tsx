import React, { Component } from 'react';
import { StyleSheet, View } from 'react-native';
import { Formik, FormikProps, FormikValues } from 'formik';
import * as yup from 'yup';
import { withTranslation, WithTranslation } from 'react-i18next';
import { FormUtils } from '@homzhub/common/src/utils/FormUtils';
import { theme } from '@homzhub/common/src/styles/theme';
import { FormTextInput, AnimatedHeader, Text, FormButton } from '@homzhub/common/src/components';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { AuthStackParamList } from '@homzhub/mobile/src/navigation/AuthStack';
import { BottomSheetListView } from '@homzhub/mobile/src/components/molecules/BottomSheetListView';

interface IVerificationState {
  phone: string;
  countryCode: string;
  isBottomSheetVisible: boolean;
}

type Props = WithTranslation & NavigationScreenProps<AuthStackParamList, ScreensKeys.MobileVerification>;

class MobileVerificationScreen extends Component<Props, IVerificationState> {
  public state = {
    phone: '',
    countryCode: '+91',
    isBottomSheetVisible: false,
  };

  public render(): React.ReactNode {
    const {
      t,
      route: {
        params: { icon, title, subTitle, buttonTitle, message },
      },
    } = this.props;
    const { isBottomSheetVisible, countryCode } = this.state;
    const formData = { ...this.state };
    return (
      <View style={styles.container}>
        <AnimatedHeader icon={icon} title={title} subTitle={subTitle} onIconPress={this.handleIconPress} />
        <View style={styles.content}>
          <Text type="small" style={styles.message}>
            {message}
          </Text>
          <Formik onSubmit={this.onSubmit} validate={FormUtils.validate(this.formSchema)} initialValues={formData}>
            {(formProps: FormikProps<FormikValues>): React.ReactElement => (
              <>
                <FormTextInput
                  formProps={formProps}
                  inputType="phone"
                  name="phone"
                  label="Phone"
                  inputPrefixText={countryCode}
                  onIconPress={this.handleDropdown}
                  placeholder={t('auth:yourNumber')}
                  helpText={t('auth:otpVerification')}
                />
                <FormButton
                  // @ts-ignore
                  onPress={formProps.handleSubmit}
                  formProps={formProps}
                  type="primary"
                  title={buttonTitle}
                  containerStyle={styles.formButtonStyle}
                />
              </>
            )}
          </Formik>
        </View>
        <BottomSheetListView
          selectedValue={countryCode}
          listTitle={t('auth:countryRegion')}
          isBottomSheetVisible={isBottomSheetVisible}
          onCloseDropDown={this.onCloseDropDown}
          onSelectItem={this.handleSelection}
        />
      </View>
    );
  }

  private onSubmit = (): void => {
    // TOD0: Add logic
  };

  private onCloseDropDown = (): void => {
    this.setState({ isBottomSheetVisible: false });
  };

  private handleIconPress = (): void => {
    const { navigation } = this.props;
    navigation.goBack();
  };

  private handleSelection = (value: string): void => {
    this.setState({ countryCode: value });
  };

  private handleDropdown = (): void => {
    const { isBottomSheetVisible } = this.state;
    this.setState({ isBottomSheetVisible: !isBottomSheetVisible });
  };

  private formSchema = (): yup.ObjectSchema<{ phone: string }> => {
    const { t } = this.props;
    return yup.object().shape({
      phone: yup.string().required(t('auth:numberRequired')),
    });
  };
}

export default withTranslation()(MobileVerificationScreen);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: theme.colors.white,
  },
  content: {
    paddingHorizontal: 24,
    paddingVertical: 16,
  },
  message: {
    color: theme.colors.darkTint3,
    marginBottom: 12,
  },
  formButtonStyle: {
    flex: 0,
    marginVertical: 30,
  },
});
