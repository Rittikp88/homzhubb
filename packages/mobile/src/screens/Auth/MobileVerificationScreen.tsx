import React, { Component } from 'react';
import { StyleSheet, View } from 'react-native';
import { Formik, FormikHelpers, FormikProps, FormikValues } from 'formik';
import * as yup from 'yup';
import { withTranslation, WithTranslation } from 'react-i18next';
import { FormUtils } from '@homzhub/common/src/utils/FormUtils';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { theme } from '@homzhub/common/src/styles/theme';
import { icons } from '@homzhub/common/src/assets/icon';
import { SocialMediaKeys } from '@homzhub/common/src/assets/constants';
import { FormTextInput, FormButton, DetailedHeader, Text } from '@homzhub/common/src/components';
import { AuthStackParamList } from '@homzhub/mobile/src/navigation/AuthStack';
import { NavigationScreenProps, OtpNavTypes, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';

interface IVerificationState {
  formData: {
    phone: string;
  };
  countryCode: string;
}

type Props = WithTranslation & NavigationScreenProps<AuthStackParamList, ScreensKeys.MobileVerification>;

export class MobileVerificationScreen extends Component<Props, IVerificationState> {
  public phone: FormTextInput | null = null;

  public state = {
    formData: {
      phone: '',
    },
    countryCode: '+91',
  };

  public render(): React.ReactNode {
    const { t } = this.props;
    const { formData, countryCode } = this.state;
    const { title, message, subTitle, buttonTitle } = this.getDisplayStrings();
    return (
      <View style={styles.container}>
        <DetailedHeader icon={icons.leftArrow} title={title} subTitle={subTitle} onIconPress={this.handleIconPress} />
        <View style={styles.content}>
          <Text type="small" style={styles.message}>
            {message}
          </Text>
          <Formik
            initialValues={{ ...formData }}
            onSubmit={this.onSubmit}
            validate={FormUtils.validate(this.formSchema)}
          >
            {(formProps: FormikProps<FormikValues>): React.ReactElement => (
              <>
                <FormTextInput
                  ref={(refs): void => {
                    this.phone = refs;
                  }}
                  formProps={formProps}
                  inputType="phone"
                  name="phone"
                  label="Phone"
                  inputPrefixText={countryCode}
                  onPhoneCodeChange={this.handlePhoneCodeChange}
                  phoneFieldDropdownText={t('auth:countryRegion')}
                  placeholder={t('yourNumber')}
                  helpText={t('otpVerification')}
                />
                <FormButton
                  // @ts-ignore
                  onPress={formProps.handleSubmit}
                  formProps={formProps}
                  type="primary"
                  title={buttonTitle}
                  containerStyle={styles.formButtonStyle}
                />
              </>
            )}
          </Formik>
        </View>
      </View>
    );
  }

  private onSubmit = (values: { phone: string }, formActions: FormikHelpers<{ phone: string }>): void => {
    formActions.setSubmitting(true);
    const {
      t,
      route: {
        params: {
          isFromLogin,
          userData: {
            user: { first_name, last_name, email },
          },
          onCallback,
        },
      },
      navigation,
    } = this.props;
    const { countryCode } = this.state;
    const { phone } = values;

    navigation.navigate(ScreensKeys.OTP, {
      ref: () => this.phone,
      type: OtpNavTypes.SocialMedia,
      title: isFromLogin ? t('loginOtp') : t('verifyNumber'),
      phone,
      countryCode,
      userData: {
        full_name: `${first_name} ${last_name}`,
        email,
        phone_number: phone,
        phone_code: countryCode,
        // TODO (Aditya 10-Jun-2020): How to solve this password issue?
        password: 'RandomPassword',
      },
      ...(onCallback && { onCallback }),
    });
  };

  private handlePhoneCodeChange = (countryCode: string): void => {
    this.setState({ countryCode });
  };

  private handleIconPress = (): void => {
    const { navigation } = this.props;
    navigation.goBack();
  };

  private getDisplayStrings = (): { title: string; subTitle: string; message: string; buttonTitle: string } => {
    const {
      t,
      route: {
        params: { userData, isFromLogin },
      },
    } = this.props;
    const {
      provider,
      user: { first_name, email },
    } = userData;

    let title = t('signUpWithFacebook');
    const messageKey = isFromLogin ? 'enterNumberForProfileForLogin' : 'enterNumberForProfileForSignUp';

    if (isFromLogin) {
      if (provider === SocialMediaKeys.Google) {
        title = t('loginWithGoogle');
      } else {
        title = t('loginWithFacebook');
      }
    }

    if (!isFromLogin && provider === SocialMediaKeys.Google) {
      title = t('signUpWithGoogle');
    }

    return {
      title,
      subTitle: email,
      buttonTitle: isFromLogin ? t('common:login') : t('common:signUp'),
      message: t(`${messageKey}`, { givenName: first_name }),
    };
  };

  private formSchema = (): yup.ObjectSchema<{ phone: string }> => {
    const { t } = this.props;
    return yup.object().shape({
      phone: yup.string().required(t('numberRequired')),
    });
  };
}

export default withTranslation(LocaleConstants.namespacesKey.auth)(MobileVerificationScreen);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: theme.colors.white,
  },
  content: {
    paddingHorizontal: 24,
    paddingVertical: 16,
  },
  message: {
    color: theme.colors.darkTint3,
    marginBottom: 12,
  },
  formButtonStyle: {
    flex: 0,
    marginVertical: 30,
  },
});
