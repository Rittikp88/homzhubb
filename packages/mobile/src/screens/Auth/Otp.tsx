import React from 'react';
import { SafeAreaView, StyleSheet, View } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';
import { AlertHelper } from '@homzhub/mobile/src/utils/AlertHelper';
import { ObjectMapper } from '@homzhub/common/src/utils/ObjectMapper';
import { UserRepository } from '@homzhub/common/src/domain/repositories/UserRepository';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { StorageService } from '@homzhub/common/src/services/storage/StorageService';
import { UserService } from '@homzhub/common/src/services/UserService';
import { UserActions } from '@homzhub/common/src/modules/user/actions';
import { UserSelector } from '@homzhub/common/src/modules/user/selectors';
import { IState } from '@homzhub/common/src/modules/interfaces';
import {
  IEmailLoginPayload,
  ILoginPayload,
  IOtpLoginPayload,
  LoginTypes,
} from '@homzhub/common/src/domain/repositories/interfaces';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { DetailedHeader, Label, OtpTimer, Text } from '@homzhub/common/src/components';
import { Loader, OtpInputs } from '@homzhub/mobile/src/components';
import { AuthStackParamList } from '@homzhub/mobile/src/navigation/AuthStack';
import { NavigationScreenProps, OtpNavTypes, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { User, IUser } from '@homzhub/common/src/domain/models/User';

interface IStateProps {
  isLoading: boolean;
}

interface IDispatchProps {
  login: (payload: ILoginPayload) => void;
  loginSuccess: (data: IUser) => void;
}

type libraryProps = NavigationScreenProps<AuthStackParamList, ScreensKeys.OTP> & WithTranslation;
type IProps = IDispatchProps & IStateProps & libraryProps;

interface IOtpState {
  error: boolean;
}

export class Otp extends React.PureComponent<IProps, IOtpState> {
  public state = {
    error: false,
  };

  public componentDidMount = async (): Promise<void> => {
    await this.fetchOtp();
  };

  public render = (): React.ReactNode => {
    const {
      t,
      isLoading,
      navigation: { goBack },
      route: { params },
    } = this.props;
    const { error } = this.state;
    const title = params?.title ?? t('verifyNumber');
    const phone = params?.phone ?? '';

    const toggleError = (): void => {
      this.toggleErrorState(false);
    };

    return (
      <SafeAreaView style={styles.screen}>
        <DetailedHeader icon={icons.leftArrow} onIconPress={goBack} headerContainerStyle={styles.headerStyle} />
        <View style={styles.container}>
          <Text type="large" textType="semiBold">
            {title}
          </Text>
          <Label style={styles.subTitle} type="large" textType="regular">
            {t('enterOTP')}
          </Label>
          <View style={styles.numberContainer}>
            <Text type="small" textType="semiBold">
              {phone}
            </Text>
            <Icon
              name={icons.noteBook}
              size={16}
              color={theme.colors.active}
              style={styles.icon}
              onPress={this.onIconPress}
              testID="icnEdit"
            />
          </View>
          <OtpInputs error={error ? t('otpError') : undefined} bubbleOtp={this.verifyOtp} toggleError={toggleError} />
          <OtpTimer onResentPress={this.fetchOtp} />
        </View>
        <Loader visible={isLoading} />
      </SafeAreaView>
    );
  };

  private onIconPress = (): void => {
    const {
      navigation,
      route: { params },
    } = this.props;

    if (params && params.ref) {
      params.ref()?.focus();
    }

    navigation.goBack();
  };

  private onVerifySuccess = async (otp?: string): Promise<void> => {
    const {
      route: {
        params: { type },
      },
    } = this.props;

    switch (type) {
      case OtpNavTypes.SignUp:
        await this.signUp();
        break;
      case OtpNavTypes.Login:
        this.loginOtp(otp ?? '');
        break;
      default:
        break;
    }
  };

  private fetchOtp = async (): Promise<void> => {
    const {
      route: {
        params: { phone, countryCode },
      },
    } = this.props;

    try {
      await UserService.fetchOtp(phone, countryCode);
    } catch (e) {
      AlertHelper.error({
        message: e.message,
      });
    }
  };

  private verifyOtp = async (otp: string): Promise<void> => {
    const {
      route: {
        params: { phone, countryCode, type },
      },
    } = this.props;

    if (type === OtpNavTypes.SocialMedia) {
      await this.socialSignUp(otp);
      return;
    }

    try {
      await UserService.verifyOtp(otp, phone, countryCode);
      await this.onVerifySuccess(otp);
    } catch (e) {
      this.toggleErrorState(true);
    }
  };

  private signUp = async (): Promise<void> => {
    const {
      login,
      route: {
        params: { userData, onCallback },
      },
    } = this.props;

    if (!userData) {
      return;
    }

    try {
      await UserRepository.signUp(userData);
      const loginData: IEmailLoginPayload = {
        action: LoginTypes.EMAIL,
        payload: {
          email: userData.email,
          password: userData.password,
        },
      };
      const loginPayload: ILoginPayload = {
        data: loginData,
        ...(onCallback && { callback: onCallback }),
      };
      login(loginPayload);
    } catch (e) {
      AlertHelper.error({ message: e.message });
    }
  };

  private socialSignUp = async (otp: string): Promise<void> => {
    const {
      loginSuccess,
      route: {
        params: { userData },
      },
    } = this.props;

    if (!userData) {
      return;
    }

    try {
      const data: User = await UserRepository.socialSignUp({
        otp,
        user_details: userData,
      });
      const serializedUser: IUser = ObjectMapper.serialize(data);
      loginSuccess(serializedUser);
      await StorageService.set<IUser>('@user', serializedUser);
    } catch (e) {
      AlertHelper.error({ message: e.message });
    }
  };

  private loginOtp = (otp: string): void => {
    const {
      login,
      route: {
        params: { phone, countryCode, onCallback },
      },
    } = this.props;

    const loginData: IOtpLoginPayload = {
      action: LoginTypes.OTP,
      payload: {
        country_code: countryCode,
        phone_number: phone,
        otp,
      },
    };

    const loginPayload: ILoginPayload = {
      data: loginData,
      callback: onCallback,
    };
    login(loginPayload);
  };

  private toggleErrorState = (error: boolean): void => {
    this.setState({ error });
  };
}

export const mapStateToProps = (state: IState): IStateProps => {
  return {
    isLoading: UserSelector.getLoadingState(state),
  };
};

export const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
  const { login, loginSuccess } = UserActions;
  return bindActionCreators(
    {
      login,
      loginSuccess,
    },
    dispatch
  );
};

export default connect<IStateProps, IDispatchProps, WithTranslation, IState>(
  mapStateToProps,
  mapDispatchToProps
)(withTranslation(LocaleConstants.namespacesKey.auth)(Otp));

const styles = StyleSheet.create({
  screen: {
    flex: 1,
    backgroundColor: theme.colors.white,
  },
  container: {
    flex: 1,
    marginTop: theme.layout.screenPaddingTop,
    marginHorizontal: theme.layout.screenPadding,
  },
  numberContainer: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  subTitle: {
    marginVertical: 8,
    color: theme.colors.darkTint3,
  },
  icon: {
    marginStart: 8,
  },
  headerStyle: {
    paddingTop: 0,
    borderBottomWidth: 0,
  },
});
