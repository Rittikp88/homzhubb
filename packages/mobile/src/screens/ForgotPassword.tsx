import React, { Component } from 'react';
import { StyleSheet, View } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { Formik, FormikProps, FormikValues } from 'formik';
import * as yup from 'yup';
import { FormUtils } from '@homzhub/common/src/utils/FormUtils';
import { AlertHelper } from '@homzhub/mobile/src/utils/AlertHelper';
import { theme } from '@homzhub/common/src/styles/theme';
import { FormButton, FormTextInput, Header, Label, Text } from '@homzhub/common/src/components';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { AuthStackParamList } from '@homzhub/mobile/src/navigation/AuthStack';

interface IForgotPasswordState {
  email: string;
}

type Props = WithTranslation & NavigationScreenProps<AuthStackParamList, ScreensKeys.ForgotPassword>;

class ForgotPassword extends Component<Props, IForgotPasswordState> {
  public state = {
    email: '',
  };

  public render(): React.ReactNode {
    const { t } = this.props;
    const formData = { ...this.state };
    return (
      <View style={styles.container}>
        <Header
          icon="left-arrow"
          title={t('auth:forgotPassword')}
          subTitle={t('auth:forgotPasswordDescription')}
          onIconPress={this.handleIconPress}
          headerContainerStyle={styles.headerContainer}
        />
        <View style={styles.content}>
          <Formik onSubmit={this.onSubmit} validate={FormUtils.validate(this.formSchema)} initialValues={formData}>
            {(formProps: FormikProps<FormikValues>): React.ReactElement => (
              <>
                <FormTextInput
                  formProps={formProps}
                  inputType="email"
                  name="email"
                  label="Email"
                  placeholder={t('auth:enterEmail')}
                />
                <FormButton
                  formProps={formProps}
                  type="primary"
                  title={t('auth:forgotPasswordRequestLink')}
                  onPress={formProps.handleSubmit}
                  containerStyle={styles.formButtonStyle}
                />
              </>
            )}
          </Formik>
          <Text type="small" textType="semiBold" style={styles.backToLoginLink} onPress={this.navigateToLogin}>
            {t('auth:backToLogin')}
          </Text>
          {/* TODO: Remove the use of Reset Password from here */}
          <Label type="large" textType="bold" style={styles.backToLoginLink} onPress={this.navigateToResetPassword}>
            Reset Password
          </Label>
        </View>
      </View>
    );
  }

  private onSubmit = (): void => {
    // TODO: Add the login on submit of Form
    AlertHelper.success({ message: 'Some message title' });
  };

  public navigateToLogin = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.EmailLogin);
  };

  public navigateToResetPassword = (): void => {
    const { navigation } = this.props;
    // TODO: Remove this navigation from here once the actual implementation is done
    navigation.navigate(ScreensKeys.ResetPassword);
  };

  public handleIconPress = (): void => {
    const { navigation } = this.props;
    navigation.goBack();
  };

  private formSchema = (): yup.ObjectSchema<{ email: string }> => {
    const { t } = this.props;
    return yup.object().shape({
      email: yup.string().required(t('auth:emailRequired')),
    });
  };
}

export default withTranslation()(ForgotPassword);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: theme.colors.white,
  },
  content: {
    paddingHorizontal: 24,
    paddingVertical: 16,
  },
  headerContainer: {
    borderBottomWidth: 0,
  },
  formButtonStyle: {
    flex: 0,
    marginVertical: 30,
  },
  backToLoginLink: {
    color: theme.colors.blue,
    alignSelf: 'center',
  },
});
