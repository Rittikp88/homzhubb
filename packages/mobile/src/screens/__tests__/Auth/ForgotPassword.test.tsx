import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import { ForgotPassword } from '@homzhub/mobile/src/screens/Auth/ForgotPassword';
import { DetailedHeader } from '@homzhub/common/src/components';

const mock = jest.fn();

describe('Forgot Password Screen', () => {
  let component: ShallowWrapper;
  let props: any;

  beforeEach(() => {
    props = {
      navigation: {
        navigate: mock,
        goBack: mock,
      },
    };
    component = shallow(
      <ForgotPassword
        {...props}
        t={(key: string): string => key}
        route={{
          params: {
            onCallback: mock,
          },
        }}
      />
    );
  });

  it('should render forgot password screen', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should navigate to login screen', () => {
    // @ts-ignore
    component.find('[testID="txtLogin"]').prop('onPress')();
    expect(mock).toHaveBeenCalled();
  });

  it('should navigate back', () => {
    // @ts-ignore
    component.find(DetailedHeader).prop('onIconPress')();
    expect(mock).toHaveBeenCalled();
  });
});
