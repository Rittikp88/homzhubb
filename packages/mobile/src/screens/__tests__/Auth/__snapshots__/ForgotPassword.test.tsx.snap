// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forgot Password Screen should render forgot password screen 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
>
  <DetailedHeader
    headerContainerStyle={
      Object {
        "borderBottomWidth": 0,
      }
    }
    icon="close"
    onIconPress={[Function]}
    subTitle="auth:forgotPasswordDescription"
    title="auth:forgotPassword"
  />
  <View
    style={
      Object {
        "paddingHorizontal": 24,
        "paddingVertical": 16,
      }
    }
  >
    <Formik
      enableReinitialize={false}
      initialValues={
        Object {
          "email": "",
        }
      }
      isInitialValid={false}
      onSubmit={[Function]}
      validate={[Function]}
      validateOnBlur={true}
      validateOnChange={true}
    >
      <Component />
    </Formik>
    <Text
      onPress={[Function]}
      style={
        Object {
          "alignSelf": "center",
          "color": "#0084F8",
        }
      }
      testID="txtLogin"
      textType="semiBold"
      type="small"
    >
      auth:backToLogin
    </Text>
  </View>
</View>
`;
